### PR TITLE
#95 extend stateful persistence context

### PR DESCRIPTION
**Pull request by gbadner**
https://github.com/hibernate/hibernate-reactive/issues/95

Supersedes https://github.com/hibernate/hibernate-reactive/pull/101.

I've rebased https://github.com/hibernate/hibernate-reactive/pull/101, resolved conflicts, and made a couple of corrections.

### CODE REVIEW

**Comment by gbadner** [comment]
The corrections were straightforward:
https://github.com/hibernate/hibernate-reactive/pull/124/commits/6f792c488b4a83c999282d48f34f91a4d0d39d2c

I made an improvement in how things are processed after loading or executing HQL:
https://github.com/hibernate/hibernate-reactive/pull/124/commits/cf0c8a309da175adcf576bde06f3c129a78ec695

Feel free to improve my improvement, or remove the commit if it doesn't look right to you.

There's no need to wait for me before merging. I'll take a look on Tuesday to see if there is anything else to do on this.

File: hibernate-rx-core/src/test/java/org/hibernate/rx/CascadeTest.java [hunk 1]
```diff
@@ -92,6 +93,8 @@ public void testCascade(TestContext context) {
 											context.assertNotNull( node );
 											context.assertEquals( node.version, 1 );
 											context.assertEquals( node.string, "Adopted");
+											context.assertTrue(Hibernate.isInitialized(node.elements));
+											context.assertFalse(Hibernate.isInitialized(node.parent));
```
Comment (gavinking) [inline hibernate-rx-core/src/test/java/org/hibernate/rx/CascadeTest.java:97]: Ah, interesting, I had never checked to see if this works.

File: hibernate-rx-core/src/main/java/org/hibernate/rx/loader/entity/impl/RxAbstractEntityLoader.java [hunk 2]
```diff
@@ -137,17 +138,13 @@ else if (e !=null ) {
...
-					}
-					return list;
-				});
+				.whenComplete( (list, e) -> persistenceContext.afterLoad() )
+				.thenCompose( list ->
+						// only initialize non-lazy collections after everything else has been refreshed
+						((RxPersistenceContextAdapter) persistenceContext ).rxInitializeNonLazyCollections()
+								.thenApply(v -> list)
+				)
+				.whenComplete( (list, e) -> persistenceContext.setDefaultReadOnly(defaultReadOnlyOrig) );
```
Comment (gavinking) [inline hibernate-rx-core/src/main/java/org/hibernate/rx/loader/entity/impl/RxAbstractEntityLoader.java:147]: Yeah, nice, I think this is better than `handle()`.

**Comment by gavinking** [comment]
Looks great to me: merge it!

**Comment by gbadner** [comment]
Thanks @gavinking!

Rebased, squashed, and merged.

**Pull request closed without merging**